ROS 2 Migration

mrceki opened the pull request:
Hi,
This PR includes my initial work on migrating the cartesian_impedance_controller to ROS 2. I've tried to follow the existing structure and maintain consistency with the original implementation.

Please feel free to create a new branch and update the base branch of the PR.

I'd also appreciate any guidance on how you usually test this controller, particularly in ROS 2. I've done some preliminary testing in sim and on real robot, but I'd like to hear how should I test it properly.

mrceki committed:
Ported Cartesian Impedance Controller from ROS 1 to ROS 2

mrceki committed:
Add custom messages ControllerConfig and ControllerState

mrceki committed:
Fix custom msg generation and enable config subscriber

mrceki committed:
New: Adds controller state publisher

mrceki committed:
Adjust default values in Cartesian Impedance Controller

mrceki committed:
Fix: Custom msg type error

mrceki committed:
Adjust default values

mrceki committed:
Fix: init nullspace !!

mrceki committed:
Fix: Parameters assignments in on_configure

mrceki committed:
Fix: Comment out test dependencies in package.xml

mrceki committed:
Fix: Remove unused dependencies from CMakeLists.txt and package.xml - Adds colcon ignore to external depends

mrceki committed:
Update build_code.yml

mrceki committed:
Rename nullspace filtering parameter to keep consistency with ROS 1 version

mrceki committed:
Cleanups

mrceki committed:
Code formatting

mrceki committed:
Update include statements for clarity

mrceki committed:
Reset wrench if apply_wrench parameter false

mrceki committed:
Add Dockerfile and update installation instructions

mrceki committed:
Update README.md

mrceki committed:
Fix robot_description handling and add to parameter config

mrceki committed:
Formating

mrceki committed:
Fix initRBDyn parameter fetch using temporary synchronous client node.

matthias-mayr commented:
Thanks a lot for this PR. A ROS 2 port is long awaited and appreciated - I have just been too busy to do it myself.

I will have a look at it and probably put it in a `ros2` branch.

qleonardolp commented:
Hi @matthias-mayr and @mrceki. I'm a PhD student working on impedance control benchmarking. I'm very interested in this ROS2 migration.

Testing the controller with a 6-Dof manipulator in my workspace, an Eigen related error crashed the controller when I activate it. Just to give some context, in my launcher the controller_manager load the `cartesian_impedance_controller` inactive. It was intentional, and up to the on_configure stage, everything was fine.

If it is of your interesting to make the source code suitable to work with 6-Dof manipulators, then I would like to suggest the following. (I'm open to support on reviewing this PR btw 😃 )

As the identity matrix dimensions are hard-coded when computing the nullspace torque vector, see [cartesian_impedance_controller.cpp#L277](https://github.com/mrceki/Cartesian-Impedance-Controller/blob/fa6e8f1360afe7923902fed886ee6280f6b78137/src/cartesian_impedance_controller.cpp#L277), my suggestion is to use the n_joints_ class member to set this matrix dimensions:

```c++
tau_nullspace << (Eigen::MatrixXd::Identity(this->n_joints_, this->n_joints_) - this->jacobian_.transpose() * jacobian_transpose_pinv) * (this->nullspace_stiffness_ * (this->q_d_nullspace_ - this->q_) - this->nullspace_damping_ * this->dq_);
```

After this change, the controller could activate without problems.

mrceki commented:
Hi @qleonardolp , thanks a lot! 

I am not sure if there is any 6-Dof torque controlled robot that you can use with this robot. I am assuming you are trying the controller UR5 in simulation but in real life this controller wont work, because universal robots are position controlled. Also to have a nullspace configuration, robot should have at least 7-Dof, so I would like to hear @matthias-mayr 's opinion on that. 

I would be happy If you can help to review PR, especially because of this is my first controller implementation.

qleonardolp commented:
> I am not sure if there is any 6-Dof torque controlled robot that you can use with this robot. I am assuming you are trying the controller UR5 in simulation but in real life this controller wont work, because universal robots are position controlled. Also to have a nullspace configuration, robot should have at least 7-Dof

Interesting, I didn't know that UR5 is position-controlled. But indeed I was testing with this model.

About the nullspace feature, I agree that it only makes sense for redundant manipulators. Otherwise it would be only necessary to deal with Jacobian singularities.

I'm kind of a ros2_control enthusiast. I'm using this framework both on my full time job and PhD. Reviewing this PR will be a win-win game. The only limitation I have is the robot hardware. I will not be able to test it on a real robot, just in Gazebo.

matthias-mayr commented:
@mrceki summer was a bit busy, but here we are finally.  
@qleonardolp If you have some time, we can look into improving this package.

I've visited the robotlab at Lund university these days and we looked into running the controller on a KUKA iiwa arm.  
I'll have a look at the PR and the package in general now.
@mrceki: I'll squash and rebase some of your changes as a start to make them easier to go through. It looks like you gave me write access to your `ros2` branch & to honor your work, I'll use it to do changes, so we can keep this PR. Your initial commits are in `ros2-pr-init` in this repo if you're interested, since the rebasing will change the history.

**Branches:**  
* ROS 1 will get `ros1`
* ROS 2 will be the new main branch of this repo due to the ROS 1 EOL

**Nullspace & DOF:**  

I'd need to reach out to the automatic control people. I'm more familiar with the coding & robotics-side. One thing that pops up in my head though is that also robots with less than 7 DOF can reach some poses with more than one joint configuration, in which case at least intuitively a nullspace configuration would make sense.  
Independent of that, this controller shall function with all possible amounts of DOF

**This controller & Robot interfaces**

@qleonardolp I see that many people stumble questions about the compatibility of compliant controllers & different robots. That's why I submitted a talk about that topic to ROSCon DE this year and in two weeks time this repo will have content about it: https://github.com/matthias-mayr/ROS2-Compliant-Arm-Control

Commit:
Fix: CMakeLists dependencies

Commit:
Chore: Change controller type and readme

Commit:
Fix: Set BUILD_TYPE to Release if not set

By default "Release" is not set, which can create issues e.g. for KUKA iiwa's and Franka Pandas if updates are missed

Commit:
Chore: Readme references main repo

Commit:
Update: Readme update with robot usage information

Commit:
Chore: Makes dependency install not fail on recompilation

For example needed in containers where the installed libraries are not kept

Commit:
Chore: Namespaces topics under controller

I prefer this as it adds semantics to the topics

Commit:
New: Enables colcon tests for ROS 2

- Enables base tests that run without ROS nodes
- Adds mock-hw "simulation" of an arm that does not move
- Runs the ROS 2 controller and observes if the output changes on inputs

Commit:
Chore: Adds pyc files to gitignore

Commit:
Docs: Updates video in Readme

Commit:
Fix: Build code with current branch of the action

matthias-mayr commented on .github/workflows/build_code.yml:
```diff
@@ -6,33 +6,39 @@ jobs:
+      - name: Checkout repository into workspace
+        uses: actions/checkout@v3
+        with:
+          ref: ros2
```
Done
